token_exchange/auth.py

```diff
@@ -0,0 +1,51 @@
+import requests
+from furl import furl
+from openforms.authentication.registry import register as registry
+from requests.auth import AuthBase
+
+from .models import TokenExchangeConfiguration
+from .signals import storage
+
+# Docs https://www.keycloak.org/docs/latest/securing_apps/#_token-exchange
+GRANT_TYPE = "urn:ietf:params:oauth:grant-type:token-exchange"
+
+
+def get_plugin_config(auth_plugin):
+    return auth_plugin.config_class.get_solo()
+
+
+class TokenAccessAuth(AuthBase):
+    def __call__(self, request):
+        config = TokenExchangeConfiguration.objects.filter(
+            service__api_root__startswith=furl(request.url).origin
+        ).first()
```

Comment:
this does not entirely honor all possible configurations - different services can be deployed on the same host/origin under different subpaths, so you need to apply a "longest-match" algorithm. This is implemented [here](https://github.com/maykinmedia/zgw-consumers/blob/main/zgw_consumers/models/services.py#L169), so I'd consider using that class method.